router/core/server.go

```diff
@@ -0,0 +1,684 @@
+		executionTransport      *http.Transport
+		baseOtelAttributes      []attribute.KeyValue
+		runtimeMetrics          *rmetric.RuntimeMetrics
+		metricStore             rmetric.Store
+		baseRouterConfigVersion string
+	}
+)
+
+func (s *server) buildMultiGraphHandler(ctx context.Context, baseMux *chi.Mux, featureFlagConfigs map[string]*nodev1.FeatureFlagRouterExecutionConfig) (http.HandlerFunc, error) {
+
+	featureFlagToMux := make(map[string]*chi.Mux, len(featureFlagConfigs))
+
+	// Build all the muxes for the feature flags in serial to avoid any race conditions
+	for featureFlagName, executionConfig := range featureFlagConfigs {
+		r, err := s.buildMux(ctx, featureFlagName, executionConfig.GetVersion(), executionConfig.GetEngineConfig(), executionConfig.Subgraphs)
+		if err != nil {
+			return nil, fmt.Errorf("failed to build mux for feature flag '%s': %w", featureFlagName, err)
+		}
+		featureFlagToMux[featureFlagName] = r
+	}
+
+	return func(w http.ResponseWriter, r *http.Request) {
+
+		// Extract the feature flag and run the corresponding mux
+		// 1. From the request header
+		// 2. From the cookie
+
+		ff := strings.TrimSpace(r.Header.Get(featureFlagHeader))
+		if ff == "" {
+			cookie, err := r.Cookie(featureFlagCookie)
+			if err == nil && cookie != nil {
+				ff = strings.TrimSpace(cookie.Value)
+			}
+		}
+
+		if mux, ok := featureFlagToMux[ff]; ok {
+			w.Header().Set(featureFlagHeader, ff)
+			mux.ServeHTTP(w, r)
+			return
+		}
+
+		// Fall back to the base composition
+		baseMux.ServeHTTP(w, r)
+	}, nil
+}
+
+func (s *server) buildMux(ctx context.Context,
+	featureFlagName string,
+	routerConfigVersion string,
+	engineConfig *nodev1.EngineConfiguration,
+	configSubgraphs []*nodev1.Subgraph) (*chi.Mux, error) {
+
+	httpRouter := chi.NewRouter()
+
+	baseOtelAttributes := append(
+		[]attribute.KeyValue{otel.WgRouterConfigVersion.String(routerConfigVersion)},
+		s.baseOtelAttributes...,
+	)
+
+	if featureFlagName != "" {
+		baseOtelAttributes = append(baseOtelAttributes, otel.WgFeatureFlag.String(featureFlagName))
+	}
+
+	subgraphs, err := configureSubgraphOverwrites(
+		engineConfig,
+		configSubgraphs,
+		s.overrideRoutingURLConfiguration,
+		s.overrides,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// We create a new execution plan cache for each operation planner which is coupled to
+	// the specific engine configuration. This is necessary because otherwise we would return invalid plans.
+	//
+	// when an execution plan was generated, which can be quite expensive, we want to cache it
+	// this means that we can hash the input and cache the generated plan
+	// the next time we get the same input, we can just return the cached plan
+	// the engine is smart enough to first do normalization and then hash the input
+	// this means that we can cache the normalized input and don't have to worry about
+	// different inputs that would generate the same execution plan
+	var planCache ExecutionPlanCache
```

Comment:
We could also have one global cache and add the FF name to the hash of the cache key.
This would give us more control over the cache size.
Currently, it's possible to configure the Cache size via the config.
With this change, the config is no longer reflecting the reality as have have many caches.
Alternatively, we could introduce a config per FF, but I think that's way more complex.
Adding the FF to the hash and one global cache is probably the easiest solution. WDYT?

Reply:
Good feedback but I'd like to avoid any further complexity for now. We can communicate that the cache size is valid for each graph "variant". I agree the total cache size would be coupled to the count of feature flags but currently a user can't have more than 3. We should revisit the whole cache implementation once this feature is more used. Even with a single cache the max memory usage is not enforced.